cpp/src/sort/segmented_sort.cu

```diff
@@ -0,0 +1,233 @@
+                      offsets_minus_one,
+                      offsets_minus_one + offsets.size(),
+                      counting_iter,
+                      counting_iter + segment_ids.size(),
+                      segment_ids.begin());
+  return std::move(segment_ids);
+}
+
+void validate_list_columns(table_view const& keys, rmm::cuda_stream_view stream)
+{
+  // check if all are list columns
+  CUDF_EXPECTS(std::all_of(keys.begin(),
+                           keys.end(),
+                           [](column_view const& col) { return col.type().id() == type_id::LIST; }),
+               "segmented_sort only supports lists columns");
+  // check if all list sizes are equal.
+  auto table_device  = table_device_view::create(keys, stream);
+  auto counting_iter = thrust::make_counting_iterator<size_type>(0);
+  CUDF_EXPECTS(
+    thrust::all_of(rmm::exec_policy(stream),
+                   counting_iter,
+                   counting_iter + keys.num_rows(),
+                   [d_keys = *table_device] __device__(size_type idx) {
+                     auto size = list_size_functor{d_keys.column(0)}(idx);
+                     return thrust::all_of(
+                       thrust::seq, d_keys.begin(), d_keys.end(), [&](auto const& d_column) {
+                         return list_size_functor{d_column}(idx) == size;
+                       });
+                   }),
+    "size of each list in a row of table should be same");
+}
+
+std::unique_ptr<column> segmented_sorted_order(table_view const& keys,
+                                               column_view const& segment_offsets,
+                                               std::vector<order> const& column_order,
+                                               std::vector<null_order> const& null_precedence,
+                                               rmm::cuda_stream_view stream,
+                                               rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(segment_offsets.type() == data_type(type_to_id<size_type>()),
+               "segment offsets should be size_type");
+  // Get segment id of each element in all segments.
+  auto segment_ids = get_segment_indices(keys.num_rows(), segment_offsets, stream);
+
+  // insert segment id before all columns.
+  std::vector<column_view> keys_with_segid;
+  keys_with_segid.reserve(keys.num_columns() + 1);
+  keys_with_segid.push_back(
+    column_view(data_type(type_to_id<size_type>()), segment_ids.size(), segment_ids.data()));
+  keys_with_segid.insert(keys_with_segid.end(), keys.begin(), keys.end());
+  auto segid_keys = table_view(keys_with_segid);
+
+  std::vector<order> child_column_order(column_order);
+  if (not column_order.empty())
+    child_column_order.insert(child_column_order.begin(), order::ASCENDING);
+  std::vector<null_order> child_null_precedence(null_precedence);
+  if (not null_precedence.empty())
+    child_null_precedence.insert(child_null_precedence.begin(), null_order::AFTER);
```

Comment:
This is ITM, but I can't think of a better way to do it. Also, `insert` to front is linear which technically could be avoided - but once again - I can't think of a cleaner way to do it.

Reply:
not clean but efficient code (avoids extra linear copy).
```suggestion
  auto prepend_default = [](auto const& vector, auto default_value) {
    if (vector.empty()) return vector;
    std::remove_cv_t<std::remove_reference_t<decltype(vector)>> pre_vector;
    pre_vector.reserve(pre_vector.size() + 1);
    pre_vector.push_back(default_value);
    pre_vector.insert(pre_vector.end(), vector.begin(), vector.end());
    return pre_vector;
  };
  auto child_column_order    = prepend_default(column_order, order::ASCENDING);
  auto child_null_precedence = prepend_default(null_precedence, null_order::AFTER);
```